core/ir/x86/decode.c

```diff
@@ -2579,6 +2580,14 @@ check_is_variable_size(opnd_t op)
 static byte *
 decode_common(dcontext_t *dcontext, byte *pc, byte *orig_pc, instr_t *instr)
 {
+    /* Synthetic ISA has its own decoder.
+     * XXX i#1684: when DR can be built with full dynamic architecture selection we won't
+     * need to pollute the decoding of other architectures with this synthetic ISA special
+     * case.
+     */
+    if (instr_get_isa_mode(instr) == DR_ISA_SYNTHETIC)
```

Comment:
Hmm, in current DR it is the dcontext's ISA mode that controls decoding, and the instr_t's ISA mode that controls encoding.  Decoding is called with a freshly empty instr_t.  Are we changing that (to work around standalone mode GLOBAL_DCONTEXT issues)?  That needs an explicit callout in the PR description and the docs: does it only apply to synth ISA or also to arm vs thumb and x86 32 vs 64?

Reply:
Even if you have this only apply to the synthetic ISA, it doesn't match the docs which say the dcontext mode is what's used for decoding (sec_isa), so a docs update would be needed regardless.

Probably what we'd want is a some "unspecified" mode value (wish we had made that 0) and then first check instr_t and if it's unspecified then use dcontext.

Reply:
I documented the exception to decoding of synthetic ISA instr in core/ir/encoding_common.h (https://github.com/DynamoRIO/dynamorio/pull/6691#discussion_r1531672652). 
I can also add it to "section ISA" in the doc.

However, I believe we can still use the ISA mode of dcontext, even for the synthetic ISA.
Strictly speaking, when we filter out traces, only the encoding needs the ISA mode to change in a tread-safe way (which we achieve using instr_t ISA mode, which now supports DR_ISA_SYNTHETIC).
Should I change the implementation to do so, so we can be consistent with how DR works?

Reply:
Yes that seems cleaner.  Leave changes to how decoding works to the issue on parallel decoding support in drdecodelib.

Reply:
Done.
All arches now check for `DR_ISA_SYNTHETIC` in `decode()` using the ISA mode of dcontext_t (not the instr_t ISA mode).